### PR TITLE
docs: Add example playbook and "readme" for confined users

### PR DESCRIPTION
**Pull request by vmojzis**
Note: Relabeling user home dirs should probably be added to https://docs.ansible.com/ansible/latest/collections/community/general/selogin_module.html

### CODE REVIEW

**Comment by richm** [comment]
Just a few minor ansible-lint issues, otherwise, lgtm
would like @bachradsusi to review for correctness

**Comment by spetrosi** [comment]
Why not include this into the main README.md?

File: examples/confined_users-playbook.yml
Comment (bachradsusi) [inline examples/confined_users-playbook.yml:62]: I'd say that `wheel` group is not necessary given that all privileged users have their own sudo config line.

File: examples/confined_users-playbook.yml
Comment (bachradsusi) [inline examples/confined_users-playbook.yml:34]: README uses "mark" and "roger"

File: examples/confined_users-playbook.yml
Comment (bachradsusi) [inline examples/confined_users-playbook.yml:44]: In "mls" we use "root" SELinux user for "root" user and `sysadm_u` for admin non-root users. I'd follow it here as well.

File: examples/confined_users-playbook.yml
Comment (bachradsusi) [reply examples/confined_users-playbook.yml:62]: I'd drop creating users completely, expect that users are already created/available and use the role to create new login mappings - https://github.com/linux-system-roles/selinux/blob/main/README.md#selinux_logins

File: examples/confined_users-playbook.yml
Comment (bachradsusi) [inline examples/confined_users-playbook.yml:107]: Would it be possible to use https://github.com/linux-system-roles/selinux/blob/main/README.md#selinux_restore_dirs instead of `restorecon` command?

Comment (vmojzis) [reply examples/confined_users-playbook.yml:107]: This command needs to take place after the login mapping is applied. So we'd need to change order of tasks in tasks/main.yml.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml:62]: It's an example, which is why I'm trying to show both ways. Anyone using the playbook can just set `privileged_users: []`.

**Comment by vmojzis** [comment]
> Why not include this into the main README.md?

The intention was to not "force" people who only want a simple way for confining users to read the whole "readme" (`confined_users-playbook.yml` uses a subset of SELinux system role's capabilities). Also, `privileged_users` variable is specific to  `confined_users-playbook.yml`.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml:44]: Ok, I guess `root` can stay "unconfined" since it's not accessible from outside. I'll replace it with a better example.

**Comment by spetrosi** [comment]
> > Why not include this into the main README.md?
> 
> The intention was to not "force" people who only want a simple way for confining users to read the whole "readme" (`confined_users-playbook.yml` uses a subset of SELinux system role's capabilities). Also, `privileged_users` variable is specific to `confined_users-playbook.yml`.

The wording of `You can greatly improve security of your systems` sounds to me like a best practice for users to do. We can also restructure the README.md to split it into "scenarios" one of which would be `Configuring confined users`, like I did in mssql role https://linux-system-roles.github.io/mssql/. @bachradsusi WDYT?
We can do this in a future PR though, after the section is merged here as a separate readme.

**Comment by bachradsusi** [comment]
Looking at https://linux-system-roles.github.io/selinux/ and https://linux-system-roles.github.io/mssql/ I guess it makes sense to merge "README-confined_users.md" into "README.md" as a new scenario.

Generally I like the idea to have confined users documented, but I think that the current approach unnecessarily mixes everything into one playbook. Also using `privileged_users` in the README is confusing for me as it's not related to the role, it's only implemented in the example playbook.

In the example playbook I'd also avoid all tasks which could be done by the role:

- instead of `shell: restorecon ...` use `selinux_restore_dirs` variable
- do not create users with assigned SELinux users, if you need to create a user, do it without SELinux part and demonstrate change of SELinux login mappings using `selinux_logins`

btw `user` module would not change SELinux user mappings if the user already exists

```
# ansible -i localhost, -c local -m user -a 'name=user_u seuser=user_u' all
localhost | CHANGED => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": true,
    "comment": "",
    "create_home": true,
    "group": 1002,
    "home": "/home/user_u",
    "name": "user_u",
    "shell": "/bin/bash",
    "state": "present",
    "system": false,
    "uid": 1002
}
# semanage login -l | grep user_u
user_u               user_u               s0                   *
# ansible -i localhost, -c local -m user -a 'name=user_u seuser=unconfined_u' all
localhost | SUCCESS => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "append": false,
    "changed": false,
    "comment": "",
    "group": 1002,
    "home": "/home/user_u",
    "move_home": false,
    "name": "user_u",
    "shell": "/bin/bash",
    "state": "present",
    "uid": 1002
}
# semanage login -l | grep user_u
user_u               user_u               s0                   *
```

**Comment by vmojzis** [comment]
> Looking at https://linux-system-roles.github.io/selinux/ and https://linux-system-roles.github.io/mssql/ I guess it makes sense to merge "README-confined_users.md" into "README.md" as a new scenario.
> 
> Generally I like the idea to have confined users documented, but I think that the current approach unnecessarily mixes everything into one playbook. Also using `privileged_users` in the README is confusing for me as it's not related to the role, it's only implemented in the example playbook.

It felt better to me to document the variable in a separate README than to leave it out. Yes, it is only used in the example playbook, but is important for anyone utilizing the example.

> 
> In the example playbook I'd also avoid all tasks which could be done by the role:

> 
> * instead of `shell: restorecon ...` use `selinux_restore_dirs` variable

`restorecon` needs to be run after the login mapping is applied. So we'd need to change order of tasks in tasks/main.yml for this to work properly.

> * do not create users with assigned SELinux users, if you need to create a user, do it without SELinux part and demonstrate change of SELinux login mappings using `selinux_logins`
> 

Why? This is an example so I feel it is important to show multiple approaches.

> btw `user` module would not change SELinux user mappings if the user already existed
> 
> ```
> # ansible -i localhost, -c local -m user -a 'name=user_u seuser=user_u' all
> localhost | CHANGED => {
>     "ansible_facts": {
>         "discovered_interpreter_python": "/usr/bin/python3"
>     },
>     "changed": true,
>     "comment": "",
>     "create_home": true,
>     "group": 1002,
>     "home": "/home/user_u",
>     "name": "user_u",
>     "shell": "/bin/bash",
>     "state": "present",
>     "system": false,
>     "uid": 1002
> }
> # semanage login -l | grep user_u
> user_u               user_u               s0                   *
> # ansible -i localhost, -c local -m user -a 'name=user_u seuser=unconfined_u' all
> localhost | SUCCESS => {
>     "ansible_facts": {
>         "discovered_interpreter_python": "/usr/bin/python3"
>     },
>     "append": false,
>     "changed": false,
>     "comment": "",
>     "group": 1002,
>     "home": "/home/user_u",
>     "move_home": false,
>     "name": "user_u",
>     "shell": "/bin/bash",
>     "state": "present",
>     "uid": 1002
> }
> # semanage login -l | grep user_u
> user_u               user_u               s0                   *
> ```

**Comment by bachradsusi** [comment]
> It felt better to me to document the variable in a separate README than to leave it out. Yes, it is only used in the example playbook, but is important for anyone utilizing the example.

It depends on what's the goal. If the goal a generic SELinux example using Ansible SELinux modules and creating users then it's probably ok. But I'd expect that README and examples in linux-system-roles would be related to roles. I understand you need to create sudoers so that administrator users could switch to sysadm_t using sudo, but I don't think it's necessary to complicate example playbook with a new list variable and use loops. It would be enough to demonstrate the change on one hardcoded user which is documented as pre-created.
 
> > In the example playbook I'd also avoid all tasks which could be done by the role:
> 
> > * instead of `shell: restorecon ...` use `selinux_restore_dirs` variable
> 
> `restorecon` needs to be run after the login mapping is applied. So we'd need to change order of tasks in tasks/main.yml for this to work properly.

You are right. It's a bug in the role and the order needs to be changed. restorecon should be the last step.


> > * do not create users with assigned SELinux users, if you need to create a user, do it without SELinux part and demonstrate change of SELinux login mappings using `selinux_logins`
> 
> Why? This is an example so I feel it is important to show multiple approaches.

It also depends on the goal. Using `user` module with `seuser` option is not related to the role.


Anyway, even though I don't agree with some parts of this PR, I know that @vmojzis is experienced SELinux engineer and so if other maintainers approve this PR I'm not going to block it.

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:49]:
```suggestion
        # Alternatively map all new users to SElinux user "staff_u" (instead of
```

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:50]:
```suggestion
        # "unconfined_u")
```

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:57]:
```suggestion
      # - {login: 'Lizzy', seuser: 'staff_u', serange: 's0', state: 'present'}
```

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:62]:
```suggestion
    # - name: Install xguest package
```

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:68]:
```suggestion
    - name: Build list of selinux logins
      set_fact:
```

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:78]: I guess this assumes you aren't using IdM or some other system that you must create users in

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:109]: why so long?  Is this because selinux has to do some sort of filesystem-wide restorecon?

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:122]:
```suggestion
      with_items: "{{ selinux_logins }}"
      changed_when: false
```

File: examples/confined_users-playbook.yml
Comment (richm) [inline examples/confined_users-playbook.yml:119]:
```suggestion
      shell: >-
        set -euo pipefail;
```

**Comment by richm** [comment]
> You are right. It's a bug in the role and the order needs to be changed. restorecon should be the last step.

Does this mean the role can't manage confined users until this is fixed?

**Comment by bachradsusi** [comment]
> Does this mean the role can't manage confined users until this is fixed?

The user part of the file context - unconfined_u, staff_u, ... - is not used for access decisions in Fedora policies so `restorecon -F` is not actually necessary. 

```
$ sestatus  | grep policy\ name
Loaded policy name:             targeted
$ id -Z
user_u:user_r:user_t:s0
$ ls -Z date
unconfined_u:object_r:user_home_t:s0 date
$ cat date
Tue Aug 29 09:59:34 AM EDT 2023
```

It could trigger security scan warnings though.

**Comment by vmojzis** [comment]
Sorry for the delay.
I dropped the use of `user` module, fixed relabeling (it is now the last task in the role and the example uses `selinux_restore_dirs`) and incorporated all the stylistic tweaks.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml:120]: This is just copied from the first example.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml]: Fixed, thanks (aside from the "-o pipefail").

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml:44]: I skipped this one since it would indent the comment from the surrounding lines.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml]: Fixed, thanks.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml]: Fixed, thanks.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml:57]: Fixed, thanks.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml]: Yes, that seemed like too complex a task to have in an example.

File: examples/confined_users-playbook.yml
Comment (vmojzis) [reply examples/confined_users-playbook.yml]: Removed (I used `selinux_restore_dirs` instead ).

File: README-confined_users.md
Comment (richm) [inline README-confined_users.md:1]:
```suggestion
# Confined users

```

File: README-confined_users.md
Comment (richm) [inline README-confined_users.md:13]:
```suggestion
### selinux_booleans

```

File: examples/confined_users-playbook.yml
Comment (richm) [reply examples/confined_users-playbook.yml:44]: ok - what do you suggest we do to make yamllint happy?

File: README-confined_users.md
Comment (vmojzis) [reply README-confined_users.md:1]: Fixed, thank you.

File: README-confined_users.md
Comment (vmojzis) [reply README-confined_users.md:14]: Fixed, thank you.

**Comment by richm** [comment]
@bachradsusi is this PR ok to merge?

**Comment by vmojzis** [comment]
Fixed some ansible-lint issues (but I cannot fix the "pipefail" and do not agree with the comment indentation on line 44).

**Comment by bachradsusi** [comment]
> @bachradsusi is this PR ok to merge?

yes

**Review by richm — APPROVED** [review]

**Pull request merged**
